peer/sync.go

```diff
@@ -0,0 +1,152 @@
+package peer
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"sync"
+
+	"github.com/renproject/aw/channel"
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+	"go.uber.org/zap"
+)
+
+type pendingContent struct {
+	// content is nil while synchronisation is happening. After synchronisation
+	// has completed, content will be set.
+	content []byte
+
+	// cond is used to wait and notify goroutines about the completion of
+	// synchronisation.
+	cond *sync.Cond
```

Comment:
Do we need a `sync.Cond` here? I don't see a case where multiple gorotuines would be waiting on the same content, since `wait` is only used inside Sync and each wait simply blocks until the context elapses/is signalled

Reply:
You could have multiple different goroutines all trying to Sync the same content concurrently. You do not want all of them to send out Pull messages, so the Cond allows all but the first to wait on existing in-flight messages.